Deprecate actor materializer

johanandren opened the pull request:
On top of #27491 and #27499, probably no reason to review before those are merged.

Refs #25559 

Deprecates the `ActorMaterializer` pulling the needed methods into `Materializer` etc. New less configurable separate factories for actor context and system for special cases.

johanandren committed:
Lots of progress on deprecating ActorMaterializer

johanandren committed:
All tests updated

patriknw commented on akka-stream/src/main/scala/akka/stream/Materializer.scala:
```diff
@@ -154,40 +190,20 @@ object Materializer {
+   * Scala API: Create a materializer whose lifecycle will be tied to the one of the passed actor. When the actor stops the materializer
+   * will stop and all streams created with it will be failed with an [[AbruptTerminationExeption]]
+   */
+  @silent("deprecated")
```
could we reverse the implementation? moving it here?

johanandren replied:
Sure

johanandren replied:
Actually, looking at it a bit, I think it is better to defer that to when removing the `ActorMaterializer` in the future instead of introducing a bunch of logic around the also deprecated `ActorMaterializerSettings` here.

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -285,6 +285,15 @@ object Sink {
    * exposes [[ActorMaterializer]] which is going to be used during materialization and
```
s/ActorMaterializer/Materializer/

probably need to search for it everywhere

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Sink.scala:
```diff
@@ -285,6 +285,15 @@ object Sink {
+   * [[Attributes]] of the [[Sink]] returned by this method.
+   */
+  @deprecated("Use 'onMaterialization' instead", "2.6.0")
   def setup[T, M](factory: BiFunction[ActorMaterializer, Attributes, Sink[T, M]]): Sink[T, CompletionStage[M]] =
```
This was added in https://github.com/akka/akka/pull/26477 and backported to 2.5, so it's right that we have to deprecate and use a different name.

Not sure `onMaterialization` is a good name though? Sounds like a callback, although the primary purpose is to be a factory?

`setupOnMaterialization` ?

johanandren replied:
I discussed it with Enno and we agreed `setup` isn't a great name to start with. I think `onMaterialization` pretty accurately describes what it does, but I agree it sounds a bit like a callback. Should it perhaps borrow from the `flatMap` terminology since it takes the output stream and "flattens" into the stream?

patriknw replied:
`flatMap` isn't well known in Java, and what are you suggesting? `onMaterializationFlatMap`?

patriknw replied:
`fromMaterializer`?

johanandren replied:
`fromMaterializer` is perfect

patriknw reviewed (commented):
first scan looking good

johanandren commented on akka-stream/src/main/scala/akka/stream/ActorMaterializer.scala:
```diff
@@ -208,45 +232,8 @@ private[akka] object ActorMaterializerHelper {
-   */
-  private[akka] def supervisor: ActorRef
-
-}
```
Yeah, I'll have to double check that.

johanandren replied:
Yeah, I'll have to double check that.

johanandren replied:
@dwijnand verified it for me, should be safe:

```
13:53:48 $ m *.scala
(press [ENTER] for the next file)

##############
# Contents of: Foo.scala
##############
package t

class Foo {
  def foo = 1
}

##############
# Contents of: Main.scala
##############
package t

object Main {
  def main(args: Array[String]): Unit = {
    println(new Foo().foo)
  }
}

13:53:50 $ scalac Foo.scala Main.scala && scala t.Main
1

13:54:07 $ vim Foo.scala

13:54:21 $ m Foo.scala
package t

class Bar {
  def foo = 2
}

class Foo extends Bar

13:54:26 $ scalac Foo.scala && scala t.Main
2
```

patriknw replied:
thanks

johanandren committed:
Mention of ActorMaterializer removed from docs

johanandren committed:
StageActor construction was never meant to be a public API

Adding a Materializer parameter proved to be tricky while keeping the
existing contstructors, but the constructor was never intended for users
to use anyway, so just filtering it out for now.

johanandren committed:
No need to use fromMaterializer here

johanandren committed:
Fixed a bunch of things, found some new problems

johanandren commented:
Will need a corresponding trait for ActorContext as I introduced for the actor systems to be able to deprecate all the `ActorMaterializer` factories for typed.

patriknw commented:
> Will need a corresponding trait for ActorContext as I introduced for the actor systems to be able to deprecate all the ActorMaterializer factories for typed.

@johanandren For Typed we don't need to deprecate. We can just change it in master since it's ApiMayChange. Maybe I misunderstand what is needed?

johanandren commented:
Ah, didn't think of that. Anyways, I'll introduce a trait that does the same for creating materializers bound to an actor.

johanandren committed:
Same materializer factory for both classic and typed ActorContext

johanandren committed:
Some more work on top

johanandren committed:
Some multi-jvm test usages of ActorMaterializer removed

johanandren committed:
Imports lost in a rebase

johanandren committed:
Formatting

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala:
```diff
@@ -63,6 +63,9 @@ private[akka] object ActorContextAdapter {
   final override val self = ActorRefAdapter(untypedContext.self)
   final override val system = ActorSystemAdapter(untypedContext.system)
+
+  private[akka] def classicActorContext = untypedContext
```
potential conflict here

patriknw commented on akka-remote/src/test/scala/akka/remote/artery/SendQueueSpec.scala:
```diff
@@ -51,7 +51,8 @@ object SendQueueSpec {
     akka.actor.serialize-messages = off
     akka.stream.materializer.debug.fuzzing-mode = on
-  """.stripMargin) with ImplicitSender {
+    akka.stream.secret-test-fuzzing-warning-disable = yep
```
wow, is `yep` valid?

johanandren replied:
Hehe, this particular setting just checks for any string, not the actual value.

patriknw commented on akka-stream/src/main/scala/akka/stream/Materializer.scala:
```diff
@@ -144,6 +145,42 @@ abstract class Materializer {
+  /**
+   * INTERNAL API
+   *
+   * FIXME this was documented as internal api but was public, can we make it internal?
```
I think it should be public. No reason to hide it, especially not now when the stream is typically run with a system.

johanandren replied:
I wonder if we should deprecate the scheduler forwarder methods as well then, since those were there when there was no public API access to the system from the materializer?

patriknw replied:
makes sense to me

patriknw commented on akka-stream/src/main/scala/akka/stream/Materializer.scala:
```diff
@@ -144,6 +145,42 @@ abstract class Materializer {
+  /**
+   * INTERNAL API
+   */
+  private[akka] def supervisor: ActorRef
```
could add some `@InternalApi` annotations while you are at it

patriknw reviewed (commented):
looking good, only some minors...

helena commented on akka-actor/src/main/mima-filters/2.5.x.backwards.excludes:
```diff
@@ -85,4 +85,6 @@ ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.DynamicAccess.c
\ No newline at end of file
+ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.actor.ActorSystem.classicSystem")
+ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.actor.AbstractActor#ActorContext.classicActorContext")
+ProblemFilters.exclude[InheritedNewAbstractMethodProblem]("akka.actor.ActorContext.classicActorContext")
```
Are we keeping a space at the EOF?

johanandren replied:
I don't think it matters for the mima excludes

helena commented on akka-docs/src/main/paradox/stream/stream-flows-and-basics.md:
```diff
@@ -411,7 +409,7 @@ Java
 This is a very useful technique if the stream is closely related to the actor, e.g. when the actor represents a user or other entity, that we continuously query using the created stream -- and it would not make sense to keep the stream alive when the actor has terminated already. The streams termination will be signalled by an "Abrupt termination exception" signaled by the stream.
 
-You may also cause an `ActorMaterializer` to shut down by explicitly calling `shutdown()` on it, resulting in abruptly terminating all of the streams it has been running then. 
+You may also cause an `Materializer` to shut down by explicitly calling `shutdown()` on it, resulting in abruptly terminating all of the streams it has been running then. 
```
```suggestion
You may also cause a `Materializer` to shut down by explicitly calling `shutdown()` on it, resulting in abruptly terminating all of the streams it has been running then. 
```

helena commented on akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/AllPersistenceIdsStage.scala:
```diff
@@ -4,31 +4,40 @@
-  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
-    new TimerGraphStageLogicWithLogging(shape) with OutHandler with Buffer[String] {
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    throw new UnsupportedOperationException("Not used")
```
Should the message be more informative?

helena replied:
Nevermind, `private[akka] class AllPersistenceIdsStage`

johanandren replied:
and the method will never be called it's just a clunky API to keep backwards compatibility.

helena commented on akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/scaladsl/LeveldbReadJournal.scala:
```diff
@@ -63,32 +70,17 @@ class LeveldbReadJournal(@unused system: ExtendedActorSystem, config: Config)
-      .mapMaterializedValue(_ => NotUsed)
-  }
+  override def currentPersistenceIds(): Source[String, NotUsed] =
+    Source.fromGraph(new AllPersistenceIdsStage(liveQuery = false, writeJournalPluginId)).named("allPersistenceIds")
```
So much cleaner now.

helena commented on akka-bench-jmh/src/main/scala/akka/stream/MaterializationBenchmark.scala:
```diff
@@ -95,7 +97,6 @@ class MaterializationBenchmark {
   import MaterializationBenchmark._
 
   implicit val system = ActorSystem("MaterializationBenchmark")
-  implicit val materializer = ActorMaterializer()
```
🎉

helena reviewed (approved):
I'm less familiar with the repercussions but this looks good, and it's pretty great. I added one minor question. Curious about https://github.com/akka/akka/pull/27538/files#r321130424. But LGTM when ready.

johanandren committed:
Feedback 1 - internal/public of methods on materializer

johanandren committed:
_a_ materializer

johanandren commented:
Woho! Conflict party 🎈 🎉 🍾

jrudolph commented on akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md:
```diff
@@ -479,19 +479,27 @@ made before finalizing the APIs. Compared to Akka 2.5.x the source incompatible
-Having a default materializer available means that most, if not all, usages of Java `ActorMaterializer.create()`
-and Scala `implicit val materializer = ActorMaterializer()` should be removed.
+The `ActorMaterializer` factories has been deprecated and replaced with a few corresponding factories in `akka.stream.Materializer`.
+New factories with per-materializer settings has not been provided but should instead be done globally through config or per stream, 
```
```suggestion
New factories with per-materializer settings have not been provided. Configuration should instead be done globally through config or per stream using attributes, 
```

jrudolph commented on akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md:
```diff
@@ -479,19 +479,27 @@ made before finalizing the APIs. Compared to Akka 2.5.x the source incompatible
 
 Details about the stream materializer can be found in [Actor Materializer Lifecycle](../stream/stream-flows-and-basics.md#actor-materializer-lifecycle)
 
+When using streams from typed the same factories and methods for creating materializers and running streams as from classic can now be used with typed. The  
```
```suggestion
When using streams from typed, the same methods can be used for running streams as from classic can now be used. The  
```

Or are there any reasons to still use factories?

jrudolph commented on akka-stream/src/main/mima-filters/2.5.x.backwards.excludes:
```diff
@@ -210,3 +211,17 @@ ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.stream.impl.FanoutProc
 ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.stream.impl.fusing.GraphInterpreterShell.this")
+# internal and only used in one test
+ProblemFilters.exclude[MissingClassProblem]("akka.stream.NoMaterializer$")
+ProblemFilters.exclude[MissingClassProblem]("akka.stream.NoMaterializer")
```
Why can't we move it to tests?

johanandren replied:
Did move it to the test it is used in, we still need to add the mima exclude

jrudolph replied:
Oops, sorry, didn't look closely enough.

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala:
```diff
@@ -600,6 +618,24 @@ private final case class SavedIslandData(
+        props.withDispatcher(context.effectiveAttributes.mandatoryAttribute[ActorAttributes.Dispatcher].dispatcher)
+      case ActorAttributes.IODispatcher.dispatcher =>
+        // this one is actually not a dispatcher but a relative config key pointing containing the actual dispatcher name
+        // FIXME go via attributes here,or something
```
Where does this method come from and how will we fix the FIXME? (follow up ticket?)

johanandren replied:
It is fixed by the indirection introduced for dispatchers earlier - when trying to load a dispatcher and that config key is a string the dispatcher lookup will just treat that string as a new config key to look up instead.

We have test coverage for this in `AttributeSpec`, I just forgot to remove this comment.

johanandren replied:
Dang, I didn't read the code carefully enough, it doesn't go through attributes, I'll fix it.

jrudolph replied:
Great, even better ;)

jrudolph commented on akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala:
```diff
@@ -1451,16 +1451,7 @@ private[stream] object Collect {
         log = logAdapter match {
           case Some(l) => l
           case _ =>
-            val mat = try ActorMaterializerHelper.downcast(materializer)
```
:+1:

jrudolph reviewed (approved):
LGTM with a few comments. This looks like a great step forward!

Can we backport `ClassicActorContextProvider` to 2.5? I guess that might be enough to make akka-http compatible with 2.5/2.6 and classic/typed ActorSystems. (We would still use the deprecated `ActorMaterializer` factories but could access a classic  ActorSystem more easily) and add `Http.apply(ClassicActorContextProvider)`.

johanandren committed:
Formatting

johanandren committed:
Some headers lost in the flood

johanandren committed:
Wait a minute... empty files reintroduced when solving rebase conflict
:/

hepin1989 commented:
have to say ,it's a good move!

prongs commented:
Hey guys, I'm using the [drunk](https://github.com/Jarlakxen/drunk/blob/master/build.sbt) library for graphql queries and its constructor is like this:

```scala
object AkkaHttpBackend {
  val ContentTypeHeader = "Content-Type"

  def apply(
    uri: Uri,
    headers: immutable.Seq[HttpHeader] = Nil,
    httpExt: Option[HttpExt] = None
  )(implicit as: ActorSystem, mat: ActorMaterializer): AkkaHttpBackend = {

    val http = httpExt.getOrElse { Http(as) }
    new AkkaHttpBackend(uri, headers, http)
  }
}
```

My question is, how do I provide an `ActorMaterializer` to it? 

Thanks to the deprecation, I do have an instance of `Materializer`, but how do I get an instance of `ActorMaterializer` from that? 

In general, how should I use a library which is expecting an instance of `ActorMaterializer` and not of a `Materializer`?

johanandren commented:
Please use the forums over at https://discuss.akka.io for questions like this @prongs, thanks!
